mldsa/native/aarch64/src/intt.S

```diff
@@ -0,0 +1,409 @@
+.endm
+
+.macro restore_vregs
+        ldp  d8,  d9, [sp, #16*0]
+        ldp d10, d11, [sp, #16*1]
+        ldp d12, d13, [sp, #16*2]
+        ldp d14, d15, [sp, #16*3]
+        add sp, sp, #(16*4)
+.endm
+
+.macro push_stack
+        save_vregs
+.endm
+
+.macro pop_stack
+        restore_vregs
+.endm
+
+.text
+.global MLD_ASM_NAMESPACE(intt_asm)
+.balign 4
+MLD_ASM_FN_SYMBOL(intt_asm)
+        push_stack
+
+        in        .req x0
+        r5678_ptr .req x1
+        r1234_ptr .req x2
+        inp       .req x3
+        count     .req x4
+        xtmp      .req x5
+
+        wtmp .req w5
+
+        data0  .req v8
+        data1  .req v9
+        data2  .req v10
+        data3  .req v11
+        data4  .req v12
+        data5  .req v13
+        data6  .req v14
+        data7  .req v15
+        data8  .req v16
+        data9  .req v17
+        data10 .req v18
+        data11 .req v19
+        data12 .req v20
+        data13 .req v21
+        data14 .req v22
+        data15 .req v23
+
+        q_data0  .req q8
+        q_data1  .req q9
+        q_data2  .req q10
+        q_data3  .req q11
+        q_data4  .req q12
+        q_data5  .req q13
+        q_data6  .req q14
+        q_data7  .req q15
+        q_data8  .req q16
+        q_data9  .req q17
+        q_data10 .req q18
+        q_data11 .req q19
+        q_data12 .req q20
+        q_data13 .req q21
+        q_data14 .req q22
+        q_data15 .req q23
+
+        root0    .req v0
+        root1    .req v1
+        root2    .req v2
+        root3    .req v3
+        root0_tw .req v4
+        root1_tw .req v5
+        root2_tw .req v6
+        root3_tw .req v7
+
+
+        q_root0    .req q0
+        q_root1    .req q1
+        q_root2    .req q2
+        q_root3    .req q3
+        q_root0_tw .req q4
+        q_root1_tw .req q5
+        q_root2_tw .req q6
+        q_root3_tw .req q7
+
+
+        tmp .req v24
+        q_tmp .req q24
+        t0  .req v25
+        t1  .req v26
+        t2  .req v27
+        t3  .req v28
+
+        modulus .req v29
+
+        // load q = 8380417
+        movz wtmp, #57345
+        movk wtmp, #127, lsl #16
+        dup modulus.4s, wtmp
+
+        mov inp, in
+
+        mov count, #16
+
+        .p2align 2
+layer5678_start:
+
+        ld4 {data0.4S, data1.4S, data2.4S, data3.4S}, [in]
+
+        load_next_roots_78 root0, root0_tw, root1, root1_tw, root2, root2_tw, r5678_ptr
+
+        gs_butterfly_v data0, data1, root1, root1_tw
+        gs_butterfly_v data2, data3, root2, root2_tw
+        gs_butterfly_v data0, data2, root0, root0_tw
+        gs_butterfly_v data1, data3, root0, root0_tw
+
+        transpose4 data0, data1, data2, data3
+
+        load_next_roots_6  root1, r1234_ptr
+        load_next_roots_56 root0, r1234_ptr
+
+        gs_butterfly data0, data1, root0, 0, 1
+        gs_butterfly data2, data3, root0, 2, 3
+        gs_butterfly data0, data2, root1, 0, 1
+        gs_butterfly data1, data3, root1, 0, 1
+
+        barrett_reduce_single data0
```

Comment:
Is this needed? We currently assume that the inputs to the iNTT are bound by Q, in which case no 32-bit overflow can happen.

Reply:
I am not quite sure where this originates from - it's taken directly from the SLOTHY paper. Possibly we wrote that to support larger inputs, but I cannot recall right now.
I agree with you that it should not be needed. I have removed it.